File: adapter/harness/trajectory_digest.py
"""
Trajectory digest — bounded, human-readable input to the supervisor
(S0 of plans/harness_trajectory_supervisor_plan.html).

``build_digest()`` is a deterministic assembly step — no LLM call. It follows the
same contract as ``escalation.SurfaceBlocker``: the output carries **no** raw
``world_model`` JSON, no ``hypothesis_set`` / ``evidence_store`` entries, and every
list field is length-capped. Only scalar and short-string projections of the run
state are pulled through.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .recovery import STRATEGY_ORDER

# Field caps — the digest is a summary, never a dump.
_MAX_GOAL = 12
_MAX_LIST = 10
_MAX_STR = 240


def _clip(text: Any, limit: int = _MAX_STR) -> str:
    s = str(text or "").strip()
    return s if len(s) <= limit else s[: limit - 1] + "…"


def _cap(items: list[Any], limit: int = _MAX_LIST) -> list[Any]:
    return list(items[:limit])


@dataclass
class TrajectoryDigest:
    """A bounded snapshot of why a run is stuck. Safe to serialise into
    HarnessRunState and to hand verbatim to an LLM."""

    goal: list[str] = field(default_factory=list)
    steps_taken: int = 0
    stall_reason: str = ""
    stall_history: list[str] = field(default_factory=list)
    strategies_tried: list[dict[str, str]] = field(default_factory=list)
    failure_classes: list[dict[str, Any]] = field(default_factory=list)
    reopened_tasks: list[str] = field(default_factory=list)
    open_contradictions: list[str] = field(default_factory=list)
    blocking_unknowns: list[str] = field(default_factory=list)
    budget_remaining: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "goal": list(self.goal),
            "steps_taken": self.steps_taken,
            "stall_reason": self.stall_reason,
            "stall_history": list(self.stall_history),
            "strategies_tried": [dict(s) for s in self.strategies_tried],
            "failure_classes": [dict(f) for f in self.failure_classes],
            "reopened_tasks": list(self.reopened_tasks),
            "open_contradictions": list(self.open_contradictions),
            "blocking_unknowns": list(self.blocking_unknowns),
            "budget_remaining": dict(self.budget_remaining),
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any] | None) -> TrajectoryDigest:
        d = d or {}
        try:
            steps = int(d.get("steps_taken", 0))
        except (TypeError, ValueError):
            steps = 0
        return cls(
            goal=_cap([_clip(g) for g in (d.get("goal") or [])], _MAX_GOAL),
            steps_taken=max(0, steps),
            stall_reason=_clip(d.get("stall_reason", "")),
            stall_history=_cap([_clip(s, 80) for s in (d.get("stall_history") or [])]),
            strategies_tried=_cap(
                [
                    {"strategy": _clip(s.get("strategy", ""), 40), "outcome": _clip(s.get("outcome", ""), 80)}
                    for s in (d.get("strategies_tried") or [])
                    if isinstance(s, dict)
                ]
            ),
            failure_classes=_cap(
                [
                    {"class": _clip(f.get("class", ""), 80), "count": int(f.get("count", 0) or 0)}
                    for f in (d.get("failure_classes") or [])
                    if isinstance(f, dict)
                ]
            ),
            reopened_tasks=_cap([_clip(t) for t in (d.get("reopened_tasks") or [])]),
            open_contradictions=_cap([_clip(c) for c in (d.get("open_contradictions") or [])]),
            blocking_unknowns=_cap([_clip(u) for u in (d.get("blocking_unknowns") or [])]),
            budget_remaining=dict(d.get("budget_remaining") or {}),
        )


def _strategies_tried(strategy_state: Any) -> list[dict[str, str]]:
    """Reconstruct the strategy sequence from switch_count + switch_triggers.

    StrategyState keeps only the current strategy, a switch count, and the trigger
    reason recorded at each switch — not a full strategy log. Walking STRATEGY_ORDER
    up to switch_count reproduces the sequence that was actually stepped through.
    """
    switch_count = int(getattr(strategy_state, "switch_count", 0) or 0)
    triggers = list(getattr(strategy_state, "switch_triggers", []))
    out: list[dict[str, str]] = []
    for i in range(min(switch_count, len(STRATEGY_ORDER))):
        out.append(
            {
                "strategy": STRATEGY_ORDER[i],
                "outcome": _clip(triggers[i], 80) if i < len(triggers) else "switched",
            }
        )
    current = getattr(strategy_state, "current_strategy", "")
    if current and (not out or out[-1]["strategy"] != current):
        out.append({"strategy": _clip(current, 40), "outcome": "current"})
    return _cap(out)


def _failure_classes(failure_diagnostics: Any) -> list[dict[str, Any]]:
    history = getattr(failure_diagnostics, "failure_history", []) if failure_diagnostics is not None else []
    counts: dict[str, int] = {}
    for entry in history:
        fc = getattr(entry, "failure_class", None) or (entry.get("failure_class") if isinstance(entry, dict) else None)
        if fc:
            counts[str(fc)] = counts.get(str(fc), 0) + 1
    ranked = sorted(counts.items(), key=lambda kv: kv[1], reverse=True)
    return [{"class": _clip(fc, 80), "count": n} for fc, n in ranked[:_MAX_LIST]]


def _open_contradictions(world_model: Any) -> list[str]:
    out: list[str] = []
    for c in getattr(world_model, "contradictions", []):
        if getattr(c, "severity", "") != "HIGH":
            continue
        desc = getattr(c, "description", "") or ""
        cid = getattr(c, "id", "?")
        if desc:
            out.append(_clip(f"{cid}: {desc}"))
        else:
            involved = ", ".join(getattr(c, "involved_belief_ids", [])[:4])
            out.append(_clip(f"{cid}: beliefs {involved}"))
    return _cap(out)


def _blocking_unknowns(world_model: Any, harness_run_state: Any) -> list[str]:
    out: list[str] = [_clip(a) for a in getattr(world_model, "assumptions", []) if str(a or "").strip()]
    pending = getattr(harness_run_state, "pending_escalation", None) if harness_run_state is not None else None
    for m in getattr(pending, "missing_info", []) or []:
        out.append(_clip(m))
    return _cap(out)


def build_digest(
    strategy_state: Any,
    failure_diagnostics: Any,
    task_graph: Any,
    world_model: Any,
    *,
    caller_state: Any | None = None,
    recovery_budget: Any | None = None,
    reopened_task_descriptions: list[str] | None = None,
    harness_run_state: Any | None = None,
) -> TrajectoryDigest:
    """Assemble a bounded TrajectoryDigest from live run state. Deterministic; no LLM."""
    success_criteria = list(getattr(caller_state, "success_criteria", []) or [])
    completion_history = list(getattr(strategy_state, "completion_history", []) or [])

    failed = [
        _clip(getattr(t, "description", "") or getattr(t, "id", ""))
        for t in getattr(task_graph, "tasks", [])
        if getattr(t, "status", "") == "FAILED"
    ]
    reopened = list(reopened_task_descriptions or []) + failed

    return TrajectoryDigest(
        goal=_cap([_clip(c) for c in success_criteria], _MAX_GOAL),
        steps_taken=len(completion_history),
        stall_reason=_clip(getattr(strategy_state, "stall_reason", "")),
        stall_history=_cap([_clip(t, 80) for t in getattr(strategy_state, "switch_triggers", [])]),
        strategies_tried=_strategies_tried(strategy_state),
        failure_classes=_failure_classes(failure_diagnostics),
        reopened_tasks=_cap([r for r in reopened if r]),
        open_contradictions=_open_contradictions(world_model),
        blocking_unknowns=_blocking_unknowns(world_model, harness_run_state),
        budget_remaining=recovery_budget.to_dict() if recovery_budget is not None else {},
    )

File: packages/harness/src/trajectory-digest.ts
// Trajectory digest — bounded, human-readable input to the supervisor
// (S0 of plans/harness_trajectory_supervisor_plan.html). TS twin of
// adapter/harness/trajectory_digest.py.
//
// buildDigest() is a deterministic assembly step — no LLM call. It follows the same
// contract as escalate()'s SurfaceBlocker: the output carries NO raw world-model
// JSON, no hypothesis-set / evidence-store entries, and every list field is
// length-capped. Only scalar / short-string projections of the run state are pulled
// through.

import { DEFAULT_STRATEGY_ORDER, type StrategyState } from './state/strategy-state.js'
import type { FailureDiagnostics } from './state/failure-diagnostics.js'
import type { TaskGraph } from './state/task-graph.js'
import type { WorldModel } from './state/world-model.js'

const MAX_GOAL = 12
const MAX_LIST = 10
const MAX_STR = 240

function clip(text: unknown, limit = MAX_STR): string {
  const s = String(text ?? '').trim()
  return s.length <= limit ? s : s.slice(0, limit - 1) + '…'
}

function cap<T>(items: T[], limit = MAX_LIST): T[] {
  return items.slice(0, limit)
}

export interface StrategyTried {
  strategy: string
  outcome: string
}
export interface FailureClassCount {
  class: string
  count: number
}

export interface TrajectoryDigestData {
  goal: string[]
  steps_taken: number
  stall_reason: string
  stall_history: string[]
  strategies_tried: StrategyTried[]
  failure_classes: FailureClassCount[]
  reopened_tasks: string[]
  open_contradictions: string[]
  blocking_unknowns: string[]
  budget_remaining: Record<string, unknown>
}

export class TrajectoryDigest {
  goal: string[]
  steps_taken: number
  stall_reason: string
  stall_history: string[]
  strategies_tried: StrategyTried[]
  failure_classes: FailureClassCount[]
  reopened_tasks: string[]
  open_contradictions: string[]
  blocking_unknowns: string[]
  budget_remaining: Record<string, unknown>

  constructor(data?: Partial<TrajectoryDigestData>) {
    this.goal = data?.goal ?? []
    this.steps_taken = data?.steps_taken ?? 0
    this.stall_reason = data?.stall_reason ?? ''
    this.stall_history = data?.stall_history ?? []
    this.strategies_tried = data?.strategies_tried ?? []
    this.failure_classes = data?.failure_classes ?? []
    this.reopened_tasks = data?.reopened_tasks ?? []
    this.open_contradictions = data?.open_contradictions ?? []
    this.blocking_unknowns = data?.blocking_unknowns ?? []
    this.budget_remaining = data?.budget_remaining ?? {}
  }

  toJSON(): TrajectoryDigestData {
    return {
      goal: [...this.goal],
      steps_taken: this.steps_taken,
      stall_reason: this.stall_reason,
      stall_history: [...this.stall_history],
      strategies_tried: this.strategies_tried.map(s => ({ ...s })),
      failure_classes: this.failure_classes.map(f => ({ ...f })),
      reopened_tasks: [...this.reopened_tasks],
      open_contradictions: [...this.open_contradictions],
      blocking_unknowns: [...this.blocking_unknowns],
      budget_remaining: { ...this.budget_remaining },
    }
  }

  static fromJSON(json: unknown): TrajectoryDigest {
    const d = (json && typeof json === 'object' ? json : {}) as Record<string, unknown>
    let steps = Number(d.steps_taken ?? 0)
    if (!Number.isFinite(steps)) steps = 0
    const asArr = (v: unknown): unknown[] => (Array.isArray(v) ? v : [])
    return new TrajectoryDigest({
      goal: cap(asArr(d.goal).map(g => clip(g)), MAX_GOAL),
      steps_taken: Math.max(0, Math.trunc(steps)),
      stall_reason: clip(d.stall_reason ?? ''),
      stall_history: cap(asArr(d.stall_history).map(s => clip(s, 80))),
      strategies_tried: cap(
        asArr(d.strategies_tried)
          .filter(s => s && typeof s === 'object')
          .map(s => {
            const o = s as Record<string, unknown>
            return { strategy: clip(o.strategy ?? '', 40), outcome: clip(o.outcome ?? '', 80) }
          }),
      ),
      failure_classes: cap(
        asArr(d.failure_classes)
          .filter(f => f && typeof f === 'object')
          .map(f => {
            const o = f as Record<string, unknown>
            let n = Number(o.count ?? 0)
            if (!Number.isFinite(n)) n = 0
            return { class: clip(o.class ?? '', 80), count: Math.trunc(n) }
          }),
      ),
      reopened_tasks: cap(asArr(d.reopened_tasks).map(t => clip(t))),
      open_contradictions: cap(asArr(d.open_contradictions).map(c => clip(c))),
      blocking_unknowns: cap(asArr(d.blocking_unknowns).map(u => clip(u))),
      budget_remaining: (d.budget_remaining && typeof d.budget_remaining === 'object'
        ? { ...(d.budget_remaining as Record<string, unknown>) }
        : {}),
    })
  }
}

function strategiesTried(ss: StrategyState): StrategyTried[] {
  const switchCount = Math.max(0, Math.trunc(ss.switch_count ?? 0))
  const triggers = ss.switch_triggers ?? []
  const out: StrategyTried[] = []
  const n = Math.min(switchCount, DEFAULT_STRATEGY_ORDER.length)
  for (let i = 0; i < n; i++) {
    out.push({
      strategy: DEFAULT_STRATEGY_ORDER[i],
      outcome: i < triggers.length ? clip(triggers[i], 80) : 'switched',
    })
  }
  const current = ss.current_strategy ?? ''
  if (current && (out.length === 0 || out[out.length - 1].strategy !== current)) {
    out.push({ strategy: clip(current, 40), outcome: 'current' })
  }
  return cap(out)
}

function failureClasses(fd: FailureDiagnostics | null | undefined): FailureClassCount[] {
  const history = fd?.failure_history ?? []
  const counts = new Map<string, number>()
  for (const entry of history) {
    const fc = (entry as { failure_class?: string })?.failure_class
    if (fc) counts.set(fc, (counts.get(fc) ?? 0) + 1)
  }
  return [...counts.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, MAX_LIST)
    .map(([fc, n]) => ({ class: clip(fc, 80), count: n }))
}

function openContradictions(wm: WorldModel): string[] {
  const out: string[] = []
  for (const c of wm.contradictions ?? []) {
    if (c.severity !== 'HIGH') continue
    const desc = c.description ?? ''
    const cid = c.id ?? '?'
    if (desc) out.push(clip(`${cid}: ${desc}`))
    else out.push(clip(`${cid}: beliefs ${(c.involved_belief_ids ?? []).slice(0, 4).join(', ')}`))
  }
  return cap(out)
}

export interface BuildDigestOptions {
  successCriteria?: string[]
  recoveryBudget?: { toJSON(): Record<string, unknown> } | Record<string, unknown> | null
  reopenedTaskDescriptions?: string[]
  missingInfo?: string[]
}

/** Assemble a bounded TrajectoryDigest from live run state. Deterministic; no LLM. */
export function buildDigest(
  strategyState: StrategyState,
  failureDiagnostics: FailureDiagnostics | null | undefined,
  taskGraph: TaskGraph,
  worldModel: WorldModel,
  opts: BuildDigestOptions = {},
): TrajectoryDigest {
  const completionHistory = strategyState.completion_history ?? []
  const failed = (taskGraph.tasks ?? [])
    .filter(t => t.status === 'FAILED')
    .map(t => clip(t.description || t.id))
  const reopened = [...(opts.reopenedTaskDescriptions ?? []), ...failed].filter(Boolean)

  let budget: Record<string, unknown> = {}
  const rb = opts.recoveryBudget
  if (rb && typeof rb === 'object') {
    budget = typeof (rb as { toJSON?: unknown }).toJSON === 'function'
      ? (rb as { toJSON(): Record<string, unknown> }).toJSON()
      : { ...(rb as Record<string, unknown>) }
  }

  return new TrajectoryDigest({
    goal: cap((opts.successCriteria ?? []).map(c => clip(c)), MAX_GOAL),
    steps_taken: completionHistory.length,
    stall_reason: clip(strategyState.stall_reason ?? ''),
    stall_history: cap((strategyState.switch_triggers ?? []).map(t => clip(t, 80))),
    strategies_tried: strategiesTried(strategyState),
    failure_classes: failureClasses(failureDiagnostics),
    reopened_tasks: cap(reopened),
    open_contradictions: openContradictions(worldModel),
    blocking_unknowns: cap([
      ...(worldModel.assumptions ?? []).filter(a => String(a ?? '').trim()).map(a => clip(a)),
      ...(opts.missingInfo ?? []).map(m => clip(m)),
    ]),
    budget_remaining: budget,
  })
}
